### PR TITLE
fix cudnn attention check

### PR DESCRIPTION
**Pull request by chengzeyi**
For CUDNN attention, besides packed QKV layout with limited support of sequence length (seq_len <= 512) and head dim requirements. Also supporting a more generic "arbitrary sequence length with flash attention" as stated in `Transformer Engine`: https://github.com/NVIDIA/TransformerEngine/blob/8e672ff0758033c348e263dbcd6a4b3578c01161/transformer_engine/common/fused_attn/fused_attn.cpp#L126

More about "fused flash attention" in CUDNN graph api: https://docs.nvidia.com/deeplearning/cudnn/developer/graph-api.html#fused-flash-attention-fprop

### CODE REVIEW

**Review by eqy — COMMENTED** [review]
Thanks for the fixes! What GPUs did you test the checks on?

**Comment by chengzeyi** [comment]
> Thanks for the fixes! What GPUs did you test the checks on?

I work with RTX 4090. And according to latest cudnn doc. This meets the tensor shape and stride requirements of "arbitrary sequence fused flash attention" rather than the quite limited "max 512 sequence length fused attention with packed QKV". However, after testing it with stable diffusion v1.5. I found that in fact cudnn's implementation of flash attention is a bit slower than the flash attention v2 in PyTorch, which is quite frustrating.

https://docs.nvidia.com/deeplearning/cudnn/developer/graph-api.html#fused-flash-attention-fprop

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 1]
```diff
@@ -286,16 +286,16 @@ auto build_graph_and_tensors(
...
+  // auto seq_q = mha_graph->tensor(fe::graph::Tensor_attributes()
+  //                                    .set_name("seq_q")
+  //                                    .set_dim({b, 1, 1, 1})
+  //                                    .set_stride({1, 1, 1, 1})
+  //                                    .set_data_type(fe::DataType_t::INT32));
+  // auto seq_kv = mha_graph->tensor(fe::graph::Tensor_attributes()
+  //                                     .set_name("seq_kv")
+  //                                     .set_dim({b, 1, 1, 1})
+  //                                     .set_stride({1, 1, 1, 1})
+  //                                     .set_data_type(fe::DataType_t::INT32));
```
Comment (eqy) [inline aten/src/ATen/native/cudnn/MHA.cpp:298]: What's the difference between the change and the commented out lines?

**Review by eqy — APPROVED** [review]

**Comment by eqy** [comment]
CC @drisspg 

I think this is a good start that we can merge ahead of the testing PR---please fix the merge conflicts @chengzeyi

**Review by drisspg — COMMENTED** [review]
Yeah, I think a rebase + fixing the commented out line section looks good, once rebased I can unblock ci/cd

File: aten/src/ATen/native/cudnn/MHA.cpp (hunk 1 above)
Comment (chengzeyi) [reply aten/src/ATen/native/cudnn/MHA.cpp:298]: > What's the difference between the change and the commented out lines?

Setting sequence length is not needed as the cudnn engine can figure it out from the shape of the input tensors.

Comment (chengzeyi) [reply aten/src/ATen/native/cudnn/MHA.cpp:298]: Seems that keeping these lines uncommented should be safe.

**Comment by chengzeyi** [comment]
@drisspg Already rebased. Note that the cudnn frontend has some extra checks about the runtime version of CUDNN. But it seems that there is currently no suitable ways to place `cudnnGetVersion()` in `sdp_utils.cpp`

**Comment by chengzeyi** [comment]
Seems that the version of cudnn can be got from `at::detail::getCUDAHooks().versionCuDNN()`

**Review by eqy — APPROVED** [review]

**Review by drisspg — APPROVED** [review]

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp [hunk 2]
```diff
@@ -301,18 +301,55 @@ bool check_all_tensors_on_device(sdp_params const& params, bool debug) {
...
-      query_lengths{params.query.sym_size(2)},
-      head_dim{params.query.sym_size(3)};
-  const bool ok = query_lengths % 64 == 0 && head_dim % 64 == 0;
-  if (!ok) {
+  const auto s_q = params.query.sym_size(2);
+  const auto s_k = params.key.sym_size(2);
+  const auto head_dim = params.query.sym_size(3);
+  long cudnn_version = at::detail::getCUDAHooks().versionCuDNN();
+  if (cudnn_version >= 90000) {
+    if (head_dim % 8 != 0 or head_dim > 256) {
```
Comment (eqy) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:309]: Looks like these need to be changed to `||`, `&&` for build issues

Comment (chengzeyi) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:309]: > Looks like these need to be changed to `||`, `&&` for build issues

Oh, it's a mistake!

**Comment by drisspg** [comment]
@chengzeyi would you mind trying to rebase?

**Comment by chengzeyi** [comment]
> @chengzeyi would you mind trying to rebase?

rebased

**Pull request closed without merging**